docs: creating steps, context.error() (#18)
zazuko/data-centric.zazuko.com: docs/workflows/how-to/fail-pipeline-gently.md

Reviewer: maybe we can add one more section covering the related problem of resource cleanup: remember we had to subscribe to events [here](https://github.com/zazuko/barnard59/blob/54f45163ee511bec85c2023f0ac11ef5f455efce/packages/cube/lib/sort.js#L15) to ensure finalization code runs also when an exception is thrown by another step

Author: Please, remind me. Did we in the end understand why that event handler was necessary? Why wasn't `store.dispose` called by the sort function enough?

Reviewer: when an exception is thrown by a later step in the pipeline, the async generator is stopped without running its `finally` block, the event handlers are the only way we could find to ensure finalization code is called. I may try to create a small repro to show the problem and how to address it

> # Fail pipeline without stopping it immediately

Reviewer: the heading is a bit long, what about something like 'Fail  pipeline gracefully'

The easiest way to stop a pipeline is to throw an exception. However, this will stop the pipeline immediately.
Consider the simple step implementation below.

```js
export default function streamFailingRandomly() {
  return async function * (source) {
    let total = 0;

    for await (const chunk of source) {
      if (Math.random() > 0.5) {
        throw new Error('Random error');
      } else {
        total += chunk.length;
        yield chunk;
      }
    }
   
    this.logger.info(`Total processed: ${total}`);
  }
}
```

When the error is thrown, the pipeline will stop immediately. No further chunks will be processed and the code following
the async loop will not be reached.

> It is quite clear from a generator-style operation implementation but can be a surprise when implementing steps using low
> level streams. In the latter case, the flush callback will also not be called.

Reviewer: ```suggestion
This is quite clear in generator-style implementations but can be a surprise when implementing steps using low-level streams. 
In the latter case, the flush callback will not be called.
```

feel free to cherry pick suggested changes

What's more, an exception thrown in one step affects other steps in the pipeline, preventing their respective flush
callbacks from being called.

:::note
`try..catch` block will not help here. When an error is thrown inside the async generator, it will be caught by the
pipeline and the processing is stopped immediately.
:::

## `this.error()` to the rescue

The solution is to use the `this.error()` method. It will instruct barnard that an error occurred but will not stop
processing. It is the responsibility of the step implementor how to handle the rest of the stream.

For the example above, the implementation could be changed to break the loop and call `this.error` instead of throwing.

```js
export default function streamFailingRandomly() {
  return async function * (source) {

>     let error = null;

Reviewer: is this line needed?

Author: Hm, no. Probably a leftover from some earlier form of this snippet
